### ISSUE TITLE
Problems building numpy with BLIS 

### ISSUE THREAD

**Issue by adamjstewart** [numpy/numpy#14180]
I'm trying to build numpy from source with BLIS as my BLAS library. Although it compiles fine, it fails its unit tests because it can't load the library.

### Reproducing code example:

BLIS is built using the following flags:
```console
$ ./configure --prefix=/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/blis-0.6.0-6hdvnljq54jw5cwmcucrmyfhv5u6ybj2 --enable-threading=none --disable-cblas --enable-blas --enable-shared --enable-static auto
```
Then, I build numpy using the following `site.cfg`:
```
[blis]
libraries = blis
library_dirs = /Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/blis-0.6.0-6hdvnljq54jw5cwmcucrmyfhv5u6ybj2/lib
rpath = /Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/blis-0.6.0-6hdvnljq54jw5cwmcucrmyfhv5u6ybj2/lib
include_dirs = /Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/blis-0.6.0-6hdvnljq54jw5cwmcucrmyfhv5u6ybj2/include
```
and the following commands:
```console
$ export NPY_BLAS_ORDER=blis
$ export NPY_LAPACK_ORDER=''
$ python -s setup.py --no-user-cfg build
$ python -s setup.py --no-user-cfg install --prefix=/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/py-numpy-1.17.0-37xrxjmnbitsruc7alrutzpvdil2xhhy --single-version-externally-managed --root=/
```

### Error message:

I see the following error message when I try to run the unit tests:
```
Traceback (most recent call last):
  File "/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/py-numpy-1.17.0-37xrxjmnbitsruc7alrutzpvdil2xhhy/lib/python3.7/site-packages/numpy/core/__init__.py", line 17, in <module>
    from . import multiarray
  File "/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/py-numpy-1.17.0-37xrxjmnbitsruc7alrutzpvdil2xhhy/lib/python3.7/site-packages/numpy/core/multiarray.py", line 14, in <module>
    from . import overrides
  File "/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/py-numpy-1.17.0-37xrxjmnbitsruc7alrutzpvdil2xhhy/lib/python3.7/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ImportError: dlopen(/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/py-numpy-1.17.0-37xrxjmnbitsruc7alrutzpvdil2xhhy/lib/python3.7/site-packages/numpy/core/_multiarray_umath.cpython-37m-darwin.so, 2): Library not loaded: libblis.2.dylib
  Referenced from: /Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/py-numpy-1.17.0-37xrxjmnbitsruc7alrutzpvdil2xhhy/lib/python3.7/site-packages/numpy/core/_multiarray_umath.cpython-37m-darwin.so
  Reason: image not found
```
There were no obvious problems during building. Here are the build logs and build environment:

- [build log](https://github.com/numpy/numpy/files/3459336/spack-build-out.txt)
- [build env](https://github.com/numpy/numpy/files/3459337/spack-build-env.txt)

It seems to me like RPATH isn't working correctly.

### Numpy/Python version information:

I'm using:

- macOS 10.14.5
- Clang 10.0.1
- Python 3.7.4
- Numpy 1.17.0
- Setuptools 41.0.1
- BLIS 0.6.0

**Comment by seberg** [comment]
I do not think it has anything to do with it, but why the `--disable-cblas`? NumPy needs the cblas ABI?

**Comment by adamjstewart** [comment]
Good question, let me try again with `--enable-cblas`. That was just the default in my package manager.

**Comment by adamjstewart** [comment]
Tried again with `--enable-cblas` but it has the same result. Same `Library not loaded: libblis.2.dylib` error during unit tests.

**Comment by charris** [comment]
Have you tried editing your `site.config` file? I keep mine in my root directory so it doesn't get deleted `~/.numpy-site.cfg`. The template is `site.cfg.example`. If there are problems with that it would be good to know.

**Comment by adamjstewart** [comment]
I write a `site.cfg` file in the tarball directory that looks like:
```
[blis]
libraries = blis
library_dirs = /Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/blis-0.6.0-6hdvnljq54jw5cwmcucrmyfhv5u6ybj2/lib
rpath = /Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/blis-0.6.0-6hdvnljq54jw5cwmcucrmyfhv5u6ybj2/lib
include_dirs = /Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/blis-0.6.0-6hdvnljq54jw5cwmcucrmyfhv5u6ybj2/include
```
I don't have any `~/.numpy-site.cfg` file that might conflict. I've successfully built numpy with MKL, OpenBLAS, Veclib/Accelerate, and Netlib-Lapack, so I think this issue might be specific to BLIS. Note that I'm setting:
```bash
export NPY_BLAS_ORDER=blis
export NPY_LAPACK_ORDER=''
```
to prevent other BLAS/LAPACK libraries from being sucked in.

**Comment by adamjstewart** [comment]
I was able to reproduce this on CentOS 7 with GCC 4.8.5. Same error message.

**Comment by rgommers** [comment]
hmm this should work. not sure if it matters whether one uses `--enable-cblas` or `BLIS_ENABLE_CBLAS`. I've marked it for the next release to at least check if we broke anything in numpy recently

**Comment by charris** [comment]
@rgommers I've pushed this off to 1.17.2. Any idea if this is a regression or not?

**Comment by rgommers** [comment]
> I've pushed this off to 1.17.2. 

that should be fine

> Any idea if this is a regression or not?

looks like a regression, but I haven't found the time to confirm that

**Comment by rth** [comment]
For the record using blis 0.6 with numpy 1.17.1 installed from conda-forge with,
```
conda create -c conda-forge -n numpy-blis numpy "blas=*=blis" python=3.7`
```
does seem to work and the tests suite mostly passes (using `pytest --pyargs numpy`) except for several failures related to the `PendingDeprecationWarning: the matrix subclass is not the recommended way to represent matrices.` warning.

**Comment by adamjstewart** [comment]
I finally found the cause of this issue! BLIS does not install its shared libraries correctly on macOS, and `install_name_tool` is needed to fix the installation. With https://github.com/spack/spack/pull/12869, I can successfully build NumPy 1.17.2 with BLIS 0.6.0 and Python 3.7.4 on macOS 10.14.6. All unit tests pass!

**Comment by rgommers** [comment]
Great! Thanks for the update @adamjstewart.
